constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/RollingThreshold.java

```diff
@@ -0,0 +1,120 @@
+package gov.nasa.jpl.aerie.constraints.tree;
+
+import gov.nasa.jpl.aerie.constraints.model.ConstraintResult;
+import gov.nasa.jpl.aerie.constraints.model.EvaluationEnvironment;
+import gov.nasa.jpl.aerie.constraints.model.LinearEquation;
+import gov.nasa.jpl.aerie.constraints.model.LinearProfile;
+import gov.nasa.jpl.aerie.constraints.model.SimulationResults;
+import gov.nasa.jpl.aerie.constraints.model.Violation;
+import gov.nasa.jpl.aerie.constraints.time.Interval;
+import gov.nasa.jpl.aerie.constraints.time.Segment;
+import gov.nasa.jpl.aerie.constraints.time.Spans;
+import gov.nasa.jpl.aerie.constraints.time.Windows;
+import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+public record RollingThreshold(Expression<Spans> spans, Expression<Duration> width, Expression<Duration> threshold, RollingThresholdAlgorithm algorithm) implements Expression<ConstraintResult> {
+
+  public enum RollingThresholdAlgorithm {
+    ExcessSpans,
+    ExcessHull,
+    DeficitSpans,
+    DeficitHull
+  }
+
+  @Override
+  public ConstraintResult evaluate(SimulationResults results, final Interval bounds, EvaluationEnvironment environment) {
+    final var width = this.width.evaluate(results, bounds, environment);
+    var spans = this.spans.evaluate(results, bounds, environment);
+
+    final Spans reportedSpans;
+    if (algorithm == RollingThresholdAlgorithm.ExcessHull || algorithm == RollingThresholdAlgorithm.ExcessSpans) {
+      reportedSpans = spans;
+    } else {
+      reportedSpans = spans.intoWindows().not().intoSpans(bounds);
+    }
+
+    final var threshold = this.threshold.evaluate(results, bounds, environment);
+
+    final var accDuration = spans.accumulatedDuration(threshold);
+    final var shiftedBack = accDuration.shiftBy(Duration.negate(width));
+
+    final var localAccDuration = shiftedBack.plus(accDuration.times(-1));
+
+    final Windows leftViolatingBounds;
+    final var violations = new ArrayList<Violation>();
+
+    final var thresholdEq = new LinearProfile(Segment.of(
+        Interval.FOREVER,
+        new LinearEquation(
+            Duration.ZERO,
+            1,
+            0
+        )
+    ));
+
+    if (algorithm == RollingThresholdAlgorithm.ExcessHull || algorithm == RollingThresholdAlgorithm.ExcessSpans) {
+      leftViolatingBounds = localAccDuration.greaterThan(thresholdEq);
```

Comment:
Why isn't the `select` part needed in the `greaterThan` case?

Reply:
This is because the end of every plan will drop below the minimum of a `deficit` thresholds when the interval starts to move off the end of the plan (because of course there are no more activities / spans). This isn't an issue for `excess` thresholds because this causes an automatic success instead of automatic violation.